R/pareto_smooth.R

```diff
@@ -0,0 +1,326 @@
+              "changing to ", S / 2, ".")
+      ndraws_tail <- S / 2
+    }
+
+    if (ndraws_tail < 5) {
+      warning("Number of tail draws cannot be less than 5. ",
+              "Changing to ", 5, ".")
+      ndraws_tail <- 5
+    }
+
+    # left tail
+    smoothed <- .pareto_smooth_tail(
+      x,
+      ndraws_tail = ndraws_tail,
+      tail = "left"
+    )
+    left_k <- smoothed$k
+
+    # right tail
+    smoothed <-.pareto_smooth_tail(
+      x = smoothed$x,
+      ndraws_tail = ndraws_tail,
+      tail = "right"
+    )
+    right_k <- smoothed$k
+
+    k <- max(left_k, right_k)
+    x <- smoothed$x
+  } else {
+    
+    smoothed <- .pareto_smooth_tail(
+      x,
+      ndraws_tail = ndraws_tail,
+      tail = tail
+    )
+    k <- smoothed$k
+    x <- smoothed$x
+  }
+
+  diags <- list(khat = k)
+
+  if (extra_diags) {
+    ext_diags <- .pareto_smooth_extra_diags(k, S)
+    diags <- c(diags, ext_diags)
+  }
+
+  if (verbose) {
+    if (!extra_diags) {
+      diags <- .pareto_smooth_extra_diags(diags$khat, length(x))
+    }
+    pareto_k_diagmsg(
+      diags = diags
+    )
+  }
+  if (return_smoothed) {
+    out <- list(x = x, diagnostics = diags)
+  } else {
+    out <- diags
+  }
+  return(out)
+}
+
+.pareto_smooth_tail <- function(x,
+                                ndraws_tail,
+                                tail = c("right", "left"),
+                                ...
+                                ) {
+
+  tail <- match.arg(tail)
+
+  S <- length(x)
+  tail_ids <- seq(S - ndraws_tail + 1, S)
+  
+  
+  if (tail == "left") {
+    x <- -x
+  }
+
+  ord <- sort.int(x, index.return = TRUE)
+  draws_tail <- ord$x[tail_ids]
+  cutoff <- ord$x[min(tail_ids) - 1] # largest value smaller than tail values
+  
+  max_tail <- max(draws_tail)
+  min_tail <- min(draws_tail)
+
+  if (ndraws_tail >= 5) {
+    ord <- sort.int(x, index.return = TRUE)
+    if (abs(max_tail - min_tail) < .Machine$double.eps / 100) {
+      warning(
+        "Can't fit generalized Pareto distribution ",
+        "because all tail values are the same.",
+        call. = FALSE
+      )
+      smoothed <- NULL
+      k <- NA
+    } else {
+      # save time not sorting since x already sorted
+      fit <- gpdfit(draws_tail - cutoff, sort_x = FALSE)
+      k <- fit$k
+      sigma <- fit$sigma
+      if (is.finite(k)) {
+        p <- (seq_len(ndraws_tail) - 0.5) / ndraws_tail
+        smoothed <- qgpd(p = p, mu = cutoff, k = k, sigma = sigma)
+      }
+    }
+  } else {
+    warning(
+      "Can't fit generalized Pareto distribution ",
+      "because ndraws_tail is less than 5.",
+      call. = FALSE
+    )
+    smoothed <- NULL
+    k <- NA
+  }
+
+  # truncate at max of raw draws
+  if (!is.null(smoothed)) {
+    smoothed[smoothed > max_tail] <- max_tail
+    x[ord$ix[tail_ids]] <- smoothed
+  }
+  
+  if (tail == "left") {
+    x <- -x
+  }
+  
+  out <- list(x = x, k = k)
+
+  return(out)
+}
+
+.pareto_smooth_extra_diags <- function(k, S, ...) {
```

Comment:
add at least one or two lines of comments to explain what these internal functions do